sheets-import/server.gs

```diff
@@ -81,72 +64,91 @@ function showSidebar() {
+ * authorization attempt. Presents an authorization results window upon
+ * completion of the API auth sequence. For additional details, see the
+ * OAuth2 Apps Script library:
+ *   https://github.com/googlesamples/apps-script-oauth2
+ * @param {Object} request results of API auth request.
  */
-function switchToReport(reportId) {
-  var config = getReportConfig(reportId);
-  activateById(config.sheetId);
-  return config;
+function authCallback(request) {
+  var template = HtmlService.createTemplateFromFile('myServiceAuthCallbackView');
+  template.user = Session.getEffectiveUser().getEmail();
+  template.isAuthorized = false;
+  template.error = null;
+  var title;
+  try {
+    var service = getService();
+    var authorized = service.handleCallback(request);
+    template.isAuthorized = authorized;
+    title = authorized ? 'Access Granted' : 'Access Denied';
+  } catch (e) {
+    template.error = e;
+    title = 'Access Error';
+  }
+  template.title = title;
+  return template.evaluate()
+      .setTitle(title)
+      .setSandboxMode(HtmlService.SandboxMode.IFRAME);
 }
 
 /**
  * Import data to the spreadsheet according to the given report
  * configuration.
- * @param {String} reportId the report identifier.
+ * @param {Object} report the report config
  * @return {Object} the (possibly updated) report configuration.
  */
-function runImport(reportId) {
+function runImport(config) {
   var ss = SpreadsheetApp.getActiveSpreadsheet();
-  var config = getReportConfig(reportId);
 
   // Acquire the sheet to place the import results in,
   // then clear and format it.
   // Update the saved config with sheet/time information.
   var sheet = activateReportSheet(config);
-  var columnIds = getColumnIds(config);
   var lastRun = new Date().toString();
-  config = updateOnImport(config, sheet, lastRun);
+  if (config.id !== undefined) {
+    // if the config has been saved before, update it in the db
+    // otherwise, just run it and see how it does
+    config = updateOnImport(config, sheet, lastRun);
+  }
 
   // Call for pages of API information to place in the sheet, one
   // page at a time.
   var pageNumber = 0;
-  var firstRow = 2;
+  var firstRow = config.clearAppend === 'clear' ? 2 : sheet.getLastRow() + 1;
   try {
     var page;
     do {
-      page = getDataPage(columnIds, pageNumber, IMPORT_PAGE_SIZE, config);
+      page = getDataPage(config.columns, pageNumber, IMPORT_PAGE_SIZE, config);
       if (page) {
+        page = padValues(page);
         sheet.getRange(firstRow, 1, page.length, page[0].length).setValues(page);
         firstRow += page.length;
         pageNumber++;
         SpreadsheetApp.flush();
       }
-    } while (page != null);
+    } while (page !== null);
   } catch (e) {
     // Ensure a new sheet Id, if created, is preserved.
-    throw ERROR_CODES.IMPORT_FAILED;
+    throw e;
   }
 
   for (var i = 1; i <= sheet.getLastColumn(); i++) {
```

Comment:
Is the ss.toast() message here still needed if you have a similar toast message popping up in the sidebar?